### ISSUE TITLE
[RFC] Adding CPU Intrinsics into codegen

### ISSUE THREAD

**Issue by cfgfung** [pytorch/pytorch#191257]
### 🚀 The feature, motivation and pitch

With the rise of agentic AI and edge computing, CPU inference performance has become increasingly critical. To unlock the full hardware potential of modern CPUs, it is often necessary to utilize low-level intrinsics. One example is [this](https://github.com/cfgfung/pytorch/blob/raymond/overlap_amx_avx_only/torch/_inductor/codegen/cpu_intrinsics/cpp_flex_attention_amx.py), which leverages the CPU's out-of-order execution engine to overlap GEMM and Softmax operations within Flex Attention 3.

### Alternatives

We are considering two approaches for placing this implementation:

**Approach 1: Place it directly under torch/_inductor/codegen**

Pros: Easy to track and review.
Cons: Tightly couples the implementation to a specific function (flex_attention).

**Approach 2: Create a new torch/_inductor/codegen/cpu_intrinsics/ directory**

Pros: Extensible for future intrinsic implementations.
Cons: Potentially over-engineered, given that flex_attention is currently the sole use case.

Recommendation: 
I favor Approach 2. With the rise of agentic AI workloads, we can anticipate a growing need for CPU-specific optimizations. 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @Chillee @drisspg @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv @jas @jansel

**Comment by cfgfung** [comment]
cc @jansel Will there be any plan to include generic CPU intrinsic in the codegen?

**Comment by aditew01** [comment]
Thanks , this is interesting. Is there any specific workflow where the usecase might make sense for an end to end flow ?
As you're recommending a singular Vectorized kernel for the targetted CPU backend for Flex attention 3? Can this be potentially implemented in oneDNN?

**Comment by cfgfung** [comment]
> Thanks , this is interesting. Is there any specific workflow where the usecase might make sense for an end to end flow ? As you're recommending a singular Vectorized kernel for the targetted CPU backend for Flex attention 3? Can this be potentially implemented in oneDNN?

There are Intel customers using Xeon CPUs for some inference workloads, which involved attention mechanism. I also expect there will be more related workloads due to agentic AI. 

oneDNN may not be the best choice for this. There are some dispatch logic, which add more instructions to the reorder buffer, in oneDNN. In that case, the out-of-order engine cannot recognize a chance to overlap the GEMM and Softmax. That is why we have to carefully write the intrinsic in order to co-issue the GEMM and Softmax.

If users seek for CPUs performance, I suppose exposing intrinsic to them is a reasonable choice. Just like [cutlass](https://github.com/cfgfung/pytorch/blob/raymond/overlap_amx_avx_only/torch/_inductor/codegen/cutlass/gemm_template.py) and [rocm](https://github.com/cfgfung/pytorch/blob/raymond/overlap_amx_avx_only/torch/_inductor/codegen/rocm/ck_tile_universal_gemm_template.py)

**Comment by aditew01** [comment]
> There are some dispatch logic, which add more instructions to the reorder buffer, in oneDNN. In that case, the out-of-order engine cannot recognize a chance to overlap the GEMM and Softmax. That is why we have to carefully write the intrinsic in order to co-issue the GEMM and Softmax.


I agree with this, for some cases it's much easier to port performance by directly plugging in intrinsics, I assume  `max-autotune` mode was meant for this, but the suggested way does seem reasonable. Happy do review the PR. 

cc: @jansel @Nicoshev

**Comment by jansel** [comment]
The proposal here is codegen only correct?  No user-facing changes?

**Comment by cfgfung** [comment]
> The proposal here is codegen only correct? No user-facing changes?

Yes. It is transparent to user.

**Comment by vpirogov** [comment]
> oneDNN may not be the best choice for this. There are some dispatch logic, which add more instructions to the reorder buffer, in oneDNN. In that case, the out-of-order engine cannot recognize a chance to overlap the GEMM and Softmax. That is why we have to carefully write the intrinsic in order to co-issue the GEMM and Softmax.

Nothing prevents integrating fused implementation into oneDNN. This works in upstream PyTorch for XPUs already.

**Comment by cfgfung** [comment]
> > oneDNN may not be the best choice for this. There are some dispatch logic, which add more instructions to the reorder buffer, in oneDNN. In that case, the out-of-order engine cannot recognize a chance to overlap the GEMM and Softmax. That is why we have to carefully write the intrinsic in order to co-issue the GEMM and Softmax.
> 
> Nothing prevents integrating fused implementation into oneDNN. This works in upstream PyTorch for XPUs already.

Could you please share the link to those fused implementations?

From a maintenance and development perspective, it’s much easier to keep the intrinsics directly within PyTorch Codegen. This lowers the barrier to entry, as contributors won't need to learn the oneDNN architecture just to modify a single kernel. They can simply update the intrinsic in the Python file and test it immediately, without having to recompile the library.

**Comment by vpirogov** [comment]
> Could you please share the link to those fused implementations?

You can find fused SDPA implementation for XPU [here](https://github.com/uxlfoundation/oneDNN/tree/main/src/gpu/intel/sdpa) and integration to PyTorch [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/xpu/Attention.cpp)

> From a maintenance and development perspective, it’s much easier to keep the intrinsics directly within PyTorch Codegen.

That's completely different argument. I wonder if PyTorch maintainers think along the same lines.

**Comment by aditew01** [comment]
IMHO oneDNN is pretty easy to integrate, it works very well for standard cases like Conv, Matmul's and the post-ops fusion works preety well. That being said, for cases like fused SDPA, considering the multiple variants of attention that are coming up, it makes it easy to have an integrated kernel in PyTorch codegen which can be tweaked and updated fairly easily. It's more lightweight to integrate and can be refactored once a standard oneDNN API is in place (which I think is not very straightforward and needs some thought on the design).

**Comment by cfgfung** [comment]
> IMHO oneDNN is pretty easy to integrate, it works very well for standard cases like Conv, Matmul's and the post-ops fusion works preety well. That being said, for cases like fused SDPA, considering the multiple variants of attention that are coming up, it makes it easy to have an integrated kernel in PyTorch codegen which can be tweaked and updated fairly easily. It's more lightweight to integrate and can be refactored once a standard oneDNN API is in place (which I think is not very straightforward and needs some thought on the design).

Agree on this. We should have the integrated kernels in PyTorch Codegen for flexibility. The LLMs are evolving fast. KIMI 3 is using a new Kimi Delta Attention mechanism. oneDNN is a great library that contains extreme performance optimizations. Once we can confirm the APIs, we can port the kernels and further optimize in oneDNN.
